gson/src/main/java/com/google/gson/internal/JavaVersion.java

```diff
@@ -26,51 +26,15 @@
   private static final int majorJavaVersion = determineMajorJavaVersion();
 
   private static int determineMajorJavaVersion() {
-    String javaVersion = System.getProperty("java.version");
-    return getMajorJavaVersion(javaVersion);
-  }
-
-  // Visible for testing only
-  static int getMajorJavaVersion(String javaVersion) {
-    int version = parseDotted(javaVersion);
-    if (version == -1) {
-      version = extractBeginningInt(javaVersion);
-    }
-    if (version == -1) {
-      return 6;  // Choose minimum supported JDK version as default
-    }
-    return version;
-  }
-
-  // Parses both legacy 1.8 style and newer 9.0.4 style 
-  private static int parseDotted(String javaVersion) {
-    try {
-      String[] parts = javaVersion.split("[._]");
-      int firstVer = Integer.parseInt(parts[0]);
-      if (firstVer == 1 && parts.length > 1) {
-        return Integer.parseInt(parts[1]);
-      } else {
-        return firstVer;
-      }
-    } catch (NumberFormatException e) {
-      return -1;
-    }
-  }
-
-  private static int extractBeginningInt(String javaVersion) {
-    try {
-      StringBuilder num = new StringBuilder();
-      for (int i = 0; i < javaVersion.length(); ++i) {
-        char c = javaVersion.charAt(i);
-        if (Character.isDigit(c)) {
-          num.append(c);
-        } else {
-          break;
-        }
-      }
-      return Integer.parseInt(num.toString());
-    } catch (NumberFormatException e) {
-      return -1;
+    int version = (int) Double.parseDouble(System.getProperty("java.class.version"));
+    switch (version) {
```

Comment:
Great... Will this work on Android?

Reply:
Looks like it: https://developer.android.com/reference/java/lang/System.html#getProperties()

Reply:
Though I would rather use Reflection-based feature detection (depends on what Gson actually is using the Java version for) rather than querying the Java version, because Android is not Java and does not have to adhere to a version of the Java or Virtual Machine Language Specification, so they can write what they want in that property.